scip_indexer/SCIPFieldResolve.cc

```diff
@@ -0,0 +1,140 @@
+    this->mixinQueue.clear();
+    for (auto mixin : query.start.data(gs)->mixins()) {
+        this->mixinQueue.push_back(mixin);
+    }
+    auto field = query.field;
+    using Data = FieldQueryResult::Data;
+    while (auto m = this->mixinQueue.try_pop_front()) {
+        auto mixin = m.value();
+        auto sym = mixin.data(gs)->findMember(gs, field);
+        if (sym.exists()) {
+            out.push_back(Data(sym));
+            continue;
+        }
+        auto it = gs.unresolvedFields.find(mixin);
+        if (it != gs.unresolvedFields.end() && it->second.contains(field)) {
+            out.push_back(Data(mixin));
+        }
+    }
+}
+
+FieldQueryResult::Data FieldResolver::findUnresolvedFieldInInheritanceChain(const core::GlobalState &gs, core::Loc loc,
+                                                                            FieldQuery query) {
+    auto start = query.start;
+    auto field = query.field;
+
+    auto fieldText = query.field.shortName(gs);
+    auto isInstanceVar = fieldText.size() >= 2 && fieldText[0] == '@' && fieldText[1] != '@';
+    auto isClassInstanceVar = isInstanceVar && start.data(gs)->isSingletonClass(gs);
+    // Class instance variables are not inherited, unlike ordinary instance
+    // variables or class variables.
+    if (isClassInstanceVar) {
+        return FieldQueryResult::Data(start);
+    }
+    auto isClassVar = fieldText.size() >= 2 && fieldText[0] == '@' && fieldText[1] == '@';
+    if (isClassVar && !start.data(gs)->isSingletonClass(gs)) {
+        // Triggered when undeclared class variables are accessed from instance methods.
+        start = start.data(gs)->lookupSingletonClass(gs);
+    }
+
+    if (gs.unresolvedFields.find(start) == gs.unresolvedFields.end() ||
+        !gs.unresolvedFields.find(start)->second.contains(field)) {
+        // Triggered by code patterns like:
+        //   # top-level
+        //   def MyClass.method
+        //     # blah
+        //   end
+        // which is not supported by Sorbet.
+        LOG_DEBUG(gs, loc,
+                  fmt::format("couldn't find field {} in class {};\n"
+                              "are you using a code pattern like def MyClass.method which is unsupported by Sorbet?",
+                              field.exists() ? field.toString(gs) : "<non-existent>",
+                              start.exists() ? start.showFullName(gs) : "<non-existent>"));
+        // As a best-effort guess, assume that the definition is
+        // in this class but we somehow missed it.
+        return FieldQueryResult::Data(start);
+    }
+
+    auto best = start;
+    auto cur = start;
+    while (cur.exists()) {
+        auto klass = cur.data(gs);
+        auto sym = klass->findMember(gs, field);
+        if (sym.exists()) { // TODO(varun): Is this early exit justified?
+                            // Maybe it is possible to hit this in multiple ancestors?
```

Comment:
Follow-up: Test this in a running Sourcegraph instance with an inheritance hierarchy involving multiple classes. In that situation, it seems like the transitive linkage should be happening through relation traversal. E.g. if `C < B < A` then `C.@f` would be defined by `B.@f` which would be defined by `A.@f`. In that case, we should check if all references to `A.@f` show up when we do Find references on `C.@f`.